torch/csrc/jit/script/compiler.cpp

```diff
@@ -0,0 +1,444 @@
+      : tree(new Def(tree_)), graph(new Graph()) {}
+
+  explicit FunctionDefinition(std::unique_ptr<Graph> graph_)
+      : tree(nullptr), graph(std::move(graph_)) {}
+
+  bool isExtern() const {
+    return tree == nullptr;
+  }
+  std::unique_ptr<Def> tree;
+  std::unique_ptr<Graph> graph;
+};
+
+} // namespace
+
+using FunctionTable = std::unordered_map<std::string, FunctionDefinition>;
+using ValueTable = std::unordered_map<std::string, Value*>;
+
+struct DefCompiler {
+  DefCompiler(FunctionDefinition& def, FunctionTable& function_table)
+      : def(def),
+        function_table(function_table) {}
+
+  // populate def->graph
+  std::vector<Value*> run() {
+    auto& tree = *def.tree;
+    for (auto input : tree.params()) {
+      auto& name = input.ident().name();
+      map(name, def.graph->addInput(name));
+    }
+    emitStatements(tree.statements());
+
+    std::vector<Value*> output_values{};
+    for (auto output : tree.returns()) {
+        auto* value = lookup(output.ident());
+        def.graph->registerOutput(value);
+        output_values.push_back(value);
+    }
+    return output_values;
+  }
+  void emitStatements(const ListView<TreeRef>& statements) {
+    for (auto stmt : statements) {
+      switch (stmt->kind()) {
+        case TK_IF:
+          emitIf(If(stmt));
+          break;
+        case TK_WHILE:
+          emitWhile(While(stmt));
+          break;
+        case TK_ASSIGN:
+          emitAssignment(Assign(stmt));
+          break;
+        case TK_GLOBAL:
+          for (auto ident : stmt->trees()) {
+            const auto& name = Ident(ident).name();
+            map(name, def.graph->addInput(name));
+          }
+          break;
+        default:
+          emitExpressionStatement(stmt);
+          break;
+      }
+    }
+  }
+  void emitIf(const If& stmt) {
+      // TODO: add support for control flow ops
+      throw ErrorReport(stmt)
+          << "Control flow is not supported yet.";
+  }
+
+  void emitWhile(const While& stmt) {
+      // TODO: add support for control flow ops
+      throw ErrorReport(stmt)
+          << "Control flow is not supported yet.";
+  }
+
+  void emitExpressionStatement(TreeRef stmt) {
+    // expression with no used outputs
+    emit(stmt);
+  }
+
+  std::vector<Value*> emitAssignment(const Assign& stmt) {
+    std::vector<Value*> outputs{stmt.lhs().size()};
+    if (stmt.reduction() != '=') {
+      if (stmt.lhs().size() != 1) {
+        throw ErrorReport(stmt)
+            << "reductions are only allow when there is a single variable "
+            << "on the left-hand side.";
+      }
+      auto lhs = stmt.lhs()[0];
+      auto expr =
+          Compound::create(stmt.reduction(), stmt.range(), {lhs, stmt.rhs()});
+      outputs = emit(expr, 1);
+    } else {
+      outputs = emit(stmt.rhs(), stmt.lhs().size());
+    }
+    int i = 0;
+    for (auto ident : stmt.lhs()) {
+      if (ident->kind() == TK_IDENT)
+        map(Ident(ident).name(), outputs.at(i));
```

Comment:
This will give a very cryptic error if you write too many entries on the lhs (something like `runtime_error: std::__vector::at`)